### ISSUE TITLE
Allow changing existing Pinot schema for certain scenarios

### ISSUE THREAD

**Issue by ksnijjer** [apache/pinot#7607]
Currently it seems like Pinot is very rigid about **only** allowing schema changes where a new column has been added. If you need to make any changes to existing columns for e.g changing column from int to long or timestamp format change, which don't affect existing data you are still forced to delete and recreate the table/schema which is very limiting and depending on amount of data that will need to be reprocessed perhaps not even feasible.

Here is a sample scenario, I created a schema where one of the datetimeSpec columns has a ingestion transform
`{
      "name": "CreationTimeMillis",
      "dataType": "LONG",
      "defaultNullValue": 1592703025,
      "transformFunction": "fromDateTime(CreateTime, \"yyyy-MM-dd'T'HH:mm:ssZ\")",
      "format": "1:MILLISECONDS:EPOCH",
      "granularity": "1:MILLISECONDS"
    }`
now apparently time format specification here is incorrect and I need to change this to

`{
      "name": "CreationTimeInMillis",
      "dataType": "LONG",
      "defaultNullValue": 1592703058,
      "transformFunction": "fromDateTime(CreateTime, 'yyyy-MM-dd''T''HH:mm:ssZ')",
      "format": "1:MILLISECONDS:EPOCH",
      "granularity": "1:MILLISECONDS"
    }`
but this is now considered as a backward incompatible change.

Can we add some additional intelligence in the logic so such changes which don't affect underlying data are allowed to go through ?

cc @mayankshriv

**Comment by klsince** [comment]
yeah, changing `defaultNullValue` should regenerate the column while reloading the segments. But need to loosen the sanity check for updateSchema rest API, which only allows to add new columns today. Basically, I think we may skip check on `defaultNullValue` [here](https://github.com/apache/pinot/blob/master/pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java#L689)

Currently, changes of `transformFunction` doesn't trigger the regeneration, and we may support it by extending the logic around [here](https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java#L234) and keeping track of the transformFunc in the segment metadata (as defined [here](https://github.com/apache/pinot/blob/master/pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/V1Constants.java#L94)) to check discrepancy.

pls lemme know your thoughts, and any interest to take a stab on them. thx!
